src/snowflake/connector/nanoarrow_cpp/ArrowIterator/CArrowChunkIterator.cpp

```diff
@@ -0,0 +1,476 @@
+//
+// Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+//
+
+#include "CArrowChunkIterator.hpp"
+#include "SnowflakeType.hpp"
+#include "IntConverter.hpp"
+#include "StringConverter.hpp"
+#include "FloatConverter.hpp"
+#include "DecimalConverter.hpp"
+#include "BinaryConverter.hpp"
+#include "BooleanConverter.hpp"
+#include "DateConverter.hpp"
+#include "TimeStampConverter.hpp"
+#include "TimeConverter.hpp"
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace sf
+{
+
+CArrowChunkIterator::CArrowChunkIterator(PyObject* context, char* arrow_bytes, int64_t arrow_bytes_size, PyObject *use_numpy)
+: CArrowIterator(arrow_bytes, arrow_bytes_size), m_latestReturnedRow(nullptr), m_context(context)
+{
+  if (py::checkPyError()) {
+    return;
+  }
+  m_currentBatchIndex = -1;
+  m_rowIndexInBatch = -1;
+  m_rowCountInBatch = 0;
+  m_latestReturnedRow.reset();
+  m_useNumpy = PyObject_IsTrue(use_numpy);
+
+  m_batchCount = m_ipcArrowArrayVec.size();
+  m_columnCount = m_batchCount > 0 ? m_ipcArrowSchema->n_children : 0;
+
+  logger->debug(__FILE__, __func__, __LINE__, "Arrow chunk info: batchCount %d, columnCount %d, use_numpy: %d", m_batchCount,
+               m_columnCount, m_useNumpy);
+}
+
+std::shared_ptr<ReturnVal> CArrowChunkIterator::next()
+{
+  m_rowIndexInBatch++;
+
+  if (m_rowIndexInBatch < m_rowCountInBatch)
+  {
+    this->createRowPyObject();
+    SF_CHECK_PYTHON_ERR()
+    return std::make_shared<ReturnVal>(m_latestReturnedRow.get(), nullptr);
+  }
+  else
+  {
+    SF_CHECK_PYTHON_ERR();
```

Comment:
this is because previously the error handling of initialization of reading arrow ipcstream was done by arrow lib in the cython layer.
however, since we pushed the initialization to the c layer, we need to detect if any error happened during that process by ourselves

Reply:
Wanna add a comment about this?

Reply:
yep, will add it

Reply:
resolved in #1758